From: John Fastabend <john.fastabend@gmail.com>
Date: Fri, 19 Jul 2019 10:29:20 -0700
Subject: bpf: sockmap, synchronize_rcu before free'ing map
Patch-mainline: v5.3-rc4
Git-commit: 2bb90e5cc90e1d09f631aeab041a9cf913a5bbe5
References: bsc#1109837

We need to have a synchronize_rcu before free'ing the sockmap because
any outstanding psock references will have a pointer to the map and
when they use this could trigger a use after free.

Fixes: 604326b41a6fb ("bpf, sockmap: convert to generic sk_msg interface")
Signed-off-by: John Fastabend <john.fastabend@gmail.com>
Signed-off-by: Daniel Borkmann <daniel@iogearbox.net>
Acked-by: Thomas Bogendoerfer <tbogendoerfer@suse.de>
---
 net/core/sock_map.c |    2 ++
 1 file changed, 2 insertions(+)

--- a/net/core/sock_map.c
+++ b/net/core/sock_map.c
@@ -252,6 +252,8 @@ static void sock_map_free(struct bpf_map
 	raw_spin_unlock_bh(&stab->lock);
 	rcu_read_unlock();
 
+	synchronize_rcu();
+
 	bpf_map_area_free(stab->sks);
 	kfree(stab);
 }
